functions/src/get-initial-state.ts

```diff
@@ -0,0 +1,42 @@
+import * as functions from "firebase-functions";
+import { Config, isEnabled, asNumber, asBoolean } from "./config";
+import {
+  computeCurrentCampaign,
+  CampaignInfo
+} from "./compute-current-campaign";
+import { firestore } from "firebase-admin";
+
+const db = firestore();
+const config = functions.config() as Config;
+const requireUniqueVote = asBoolean(
+  config.features.voting_campaigns.require_unique_vote
+);
+
+export const getInitialState = functions.https.onCall(
+  async (
+    _,
+    context
+  ): Promise<{
+    campaign: string | null;
+    alreadyVoted: boolean;
+  }> => {
+    const voteDate = new Date();
+    const campaign = computeCurrentCampaign(voteDate, {
+      enabled: isEnabled(config.features.voting_campaigns),
+      startOn: asNumber(config.features.voting_campaigns.start_on),
+      endOn: asNumber(config.features.voting_campaigns.end_on)
+    });
+    const voterEmail: string = context.auth!.token.email;
+
+    const vote = await db
+      .collection("vote")
+      .where("email", "==", voterEmail)
+      .where("campaign", "==", campaign.open ? campaign.id : "")
```

Comment:
I think that for perf reasons we should avoid the firebase query completely if there is no campaign open or if unique votes are not required.

Reply:
Suggestion: implement this with an early return.